services/core/java/com/android/server/mperspective/PerspectiveService.java

```diff
@@ -364,24 +391,25 @@ public void handleMessage(Message msg) {
                 case MSG_UPDATE_DESKTOP_STATE:
                     updateDesktopState();
                     break;
+                case MSG_DISABLE_DESKTOP_INTERACTIVE:
+                    updateDesktopInteractiveStateLocked(false);
```

Comment:
We need to run this in a `synchronized (mLock)` block to prevent concurrency races. You can create a method in this handler class like `updateDesktopInteractiveStateInternal()` like we do for `startDesktopPerspectiveInternal()` and use a synchronized block there.